Make solo half of split keyboards (more) usable.

firetech opened the pull request:
## Description

Using only one half of a split keyboard (that's using the split_common framework to communicate) is not a great experience, since several read timeouts per scan cycle cause an unusably slow scan rate.

This change blocks all split communication attempts for 500 ms (configurable) after a number of consecutive errors, causing the scan rate to become way more usable overall, but might need some tweaking to work fully on all keyboards. One read timeout still needs to occur after the 500 ms has passed (to check if the connection is back), and if that timeout isn't low enough, some (but not all) scan cycles may still be too slow.

I've only tested this on Ergodox Infinity (due to a severe lack of other keyboards). The default `SERIAL_USART_TIMEOUT` of 100 ms was a bit too much, and some testing showed that timeouts as low as 10 ms work with even the slowest defined baud rate for the serial_usart driver (19200 baud, with `SELECT_SOFT_SERIAL_SPEED` set to 5), so I've lowered the default timeout to 20 ms, which should be good for most cases, with a good margin.

~I have no means of testing split with `USE_I2C`, but the default `SLAVE_I2C_TIMEOUT` (also 100 ms) should probably also be lowered, if that can be done safely (if not, this change should at least _improve_ how a solo half with I2C behaves, even if it may not be fully usable).~ Doesn't seem to be working with `USE_I2C` since I2C communication attempts seem to hang the mcu completely while the split is disconnected, but this change should at least not break anything for keyboards using that.

My matrix scan rate (according to `DEBUG_MATRIX_SCAN_RATE`) is around 900 with the halves connected, and around 1800 with a solo half. Not sure how this scales on other keyboards.

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Could potentially be a solution to #13352

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented:
Testing this,it causes the secondary side to be unstable. unfortunately.

firetech commented:
> Testing this,it causes the secondary side to be unstable. unfortunately.

Hmm. What board? I²C or serial communication?

I could make a change so only X consecutive errors make the connection drop, might solve some corner cases.

drashna commented:
using my tractyl manuform.  So at90usb1286 (teensy++ 2.0's), using serial communication.

drashna commented:
That's a lot better!

firetech commented:
@drashna That was quick. I was still in the middle of writing a comment about the new commit. 😛 

I realized the `transaction_handler_master` function (just below the wrapper I added) gives each transaction 10 attempts before giving up, so seeing the connection as down on the first error was maybe a bit counterproductive... Therefore, I changed so the connection is only seen as down after 25 consecutive errors (a value that can easily be changed if needed, my thoughts: "two completely failed transactions, and a bit more"). Should fix instabilities if the keyboard usually behaves as it should in normal cases.

drashna commented:
Yup, had the PR open and saw the update.  And was testing some settings for it.  

And I know that my keyboard does get a lot of random drops, so checking sequential failures is probably the best way to go. 

Also, should probably add a note about how it's handled, so that it's clear what it's doing, basically.  (eg, that each transaction attempts 10 times until failing, so 25 is 2 full failures and a partial).  Or have the setting just be for full failures.

Either way, I can type without missed events now.  Will test on other boards (eg proton c based) a bit later.

firetech commented:
Added some comments to the new defines, and increased the max number of errors from 25 to 40.

The new defines should probably be added to the documentation as well, though...

drashna commented:
> The new defines should probably be added to the documentation as well, though...

Yup. Please.

KarlK90 commented on quantum/split_common/transactions.c:
```diff
@@ -53,6 +67,36 @@ void slave_rpc_exec_callback(uint8_t initiator2target_buffer_size, const void *i
+    if (connection_errors >= SPLIT_MAX_CONNECTION_ERRORS && timer_elapsed(connection_check_timer) < SPLIT_CONNECTION_CHECK_TIMEOUT) {
+        return false;
+    }
+    connection_check_timer = timer_read();
```
(I only mean line 81 the timer_read())
You could move this into the error disconnect condition in line 88. As I read the code, it indicates the time of last successful connection and as such the timer_read() doesn't need to be in the happy path, saving some cycles. Moving it there would start counting from the time it really failed, which is okay?

firetech replied:
Hmm. True.

KarlK90 reviewed (commented):
Very cool! I have no experience with bitbang serial, but the current timeout default for hardware backed usarts could be greatly reduced. My board is running with a timeout of 5ms and that is completely fine.

KarlK90 commented:
There is also the "old" disconnect logic in [matrix.c](https://github.com/qmk/qmk_firmware/blob/develop/quantum/matrix.c#L313-L326) I think this could be thrown out completely in favor of this. The resetting of the secondary matrix could be done in `transactions.c` as well. Maybe the error_counter or some global boolean connection boolean could be checked in the matrix transport task.

daskygit commented:
> 
> 
> There is also the "old" disconnect logic in [matrix.c](https://github.com/qmk/qmk_firmware/blob/develop/quantum/matrix.c#L313-L326) I think this could be thrown out completely in favor of this. The resetting of the secondary matrix could be done in `transactions.c` as well. Maybe the error_counter or some global connection boolean could be checked in the matrix transport task.

This old disconnect logic has caused quite a few problems on master recently. It seems to be in combination with rgb matrix mostly but it causes repeated keys. So something that only resets the matrix after an extended time would be good in my books.

firetech commented:
The matrix.c logic could probably be improved, yes. Not really sure exactly where to take it, though... Happy to take suggestions. :)

That said, cleaning that up could probably be a separate PR. It should continue to work as is in combination with this change.

firetech commented:
I did some (somewhat minor) changes:
* Rebased on latest develop.
* Moved some code from transactions.c to transport.c. Thought transport.c was a better place for the `transport_transaction` wrapper.
* Added a new `is_transport_connected` function (main reason for the move above what thatI thought that _this_ function definitely should be in transport.c).
* Used the `is_transport_connected` function in matrix.c (for resetting the slave matrix), like @KarlK90 suggested above.

firetech commented:
Oops, wasn't aware that re-requesting a review from drashna would remove the collaborators request. 😳

tzarc commented:
Seems like it's mostly functional, however scan rate dropped from 5000/sec to 120/sec.
That said, I'm not sure this is the right abstraction point.

This was my first attempt at doing so -- scan rate goes from 5000 to 20000, with the same sort of limits:
https://github.com/tzarc/qmk_firmware/commit/2c3e1da43ed457cadce8f64b5073f2aeaba620b3

The relevant changes are situated in `quantum/split_common/transport.c`.

Obviously it was incomplete, but it does highlight where I thought the abstraction was likely best placed. I'm not saying you need to change your PR to match, but it did seem like it was a bit cleaner.

firetech commented:
@tzarc Huh. I didn't expect that much difference! 😮

The main reason I implemented it where I did was to also affect RPC transactions, but your method plus some logic in `transaction_rpc_exec` (return false if transport isn't connected) might be the way to go then... I'll do some experiments.

firetech commented:
Moved the disconnection logic to `transport_master`, as @tzarc suggested. Caused scan rate while disconnected to jump to ~1800 (i.e. double that of connected) for me. 🙂

drashna commented:
I'll have to retest this, because I came to say that the changes .... were not good and killed the scan rate for me.   So looks like it's addressed already, and works much better

firetech commented:
> killed the scan rate for me.

With both halves or just a solo half? The effect on scanning rate while connected should have been negligible (granted, even more so with the latest revision).

drashna commented:
> > killed the scan rate for me.
> 
> With both halves or just a solo half? The effect on scanning rate while connected should have been negligible (granted, even more so with the latest revision).

when only one side was connected.  Pulling the updated code, and it's much better, and very usable.  So, it's good now.

drashna commented:
Though, it should be noted that it does take a bit for it to settle down when it doesn't have the secondary connection, but that should be configurable, at least.

firetech commented:
> Though, it should be noted that it does take a bit for it to settle down when it doesn't have the secondary connection, but that should be configurable, at least.

Yeah, it takes about 10 seconds on my Ergodox Infinity with the default settings. Lowering `SPLIT_MAX_CONNECTION_ERRORS` brings that time down, but also increases the risk of false connection losses (however, to be fair, the current default of 10 would be roughly equivalent to 100 with my old implementation, and that had 40 as default).

Another thing, for this to be fully useful on keyboards using the serial_usart driver (ARM) or `USE_I2C`, the default `SERIAL_USART_TIMEOUT` and `SLAVE_I2C_TIMEOUT` (both are currently 100 ms) should be reduced, otherwise one scan cycle every 500 ms (`SPLIT_CONNECTION_CHECK_TIMEOUT`) will take >100 ms, a bit too long for most typists.

serial_usart works fine on my Ergodox Infinity with `SERIAL_USART_TIMEOUT` set to 50 ms, but it could probably be even lower without much issues. KarlK90 mentioned [above](https://github.com/qmk/qmk_firmware/pull/13523#pullrequestreview-704484790) that 5 ms is "completely fine", and rgbkb/mun (the only keyboard except Ergodox Infinity to not use the default on current develop, it seems) is using that value. I'm therefore considering to lower the default `SERIAL_USART_TIMEOUT` in this PR while I'm at it. Would 5 ms be too low?

I have no experience with (or way to test) split with `USE_I2C`, so I don't know what timeout values would be sensible in that case, or if `SLAVE_I2C_TIMEOUT` can even be lowered safely at all.

drashna commented:
Unfortunately, I don't have an i2c based split common board.  The only one I have, has been shipped out, so I can't test.

But yeah, it's 100ms for i2c timeout. and .. it's not configurable for ARM currently, IIRC.

tzarc commented:
I'll give it a go on my Iris later today, hopefully.

firetech commented:
I lowered the default `SERIAL_USART_TIMEOUT` to 20 ms after some testing. 10 ms seemed safe on my Ergodox Infinity with even the lowest defined baud rate of 19200 (`SELECT_SOFT_SERIAL_SPEED` = 5), so 20 ms should be safe for most cases, with a good margin. (For 19200 baud, 5 ms was too low.)

As mentioned in the original description, I somewhat mistakenly included the line `#define SERIAL_USART_TIMEOUT 50` in ergodox_infinity/config.h in my previous PR #13481. Would it be OK to remove that in this PR or should I create a new PR for removing that line? Doesn't really affect anything noticeably, it's mainly cleanup.

tzarc commented:
> 
> 
> I'll give it a go on my Iris later today, hopefully.

Completely non-functional over I2C, sadly.

firetech commented:
> Completely non-functional over I2C, sadly.

☹️ 

> * Disconnecting the split while it's running results in the master completely stopping -- doesn't even report matrix scan rate over console.

It it the same if you start with the split disconnected?

tzarc commented:
Yeah, basically dead on connect -- USB fails to enumerate.

firetech commented:
Sounds like the I2C driver is missing some timeout (or unable to timeout some operations)... I guess the behavior is basically the same without this patch?

tzarc commented:
Will test in a bit. In the meantime, `transport.c` needs:
```c
#include <timer.h>
```
...added to the list of includes for I2C-based split.

tzarc commented:
Yeah, identical behaviour with `develop`.

firetech commented:
I might recruit a friend to help me jerry-rig together some pro micros with I2C to try to figure out what is going on there (i.e. what is hanging). In the meantime, can we live without I2C support for this feature, given that it doesn't seem to break anything in that case?

daskygit commented:
Could this be made opt-in? I know the firmware size increase isn't much but it's just something else AVR users have to contend with. 

From a quick multibuild this breaks a few keymaps I only tested default/via. It's mostly oversize but a couple aren't.
```
#oversize
helix/pico/sc:default
keebio/kbo5000/rev1:via
keebio/quefrency/rev2:via
keebio/quefrency/rev3:via
keebio/sinc/rev1:via
keebio/sinc/rev2:via

#broken
handwired/lagrange:default
whale/sk/v3:default
```

firetech commented:
Making it opt-in shouldn't be much of an issue, if desired. It should be more or less opt-out as it is currently, as setting `SPLIT_MAX_CONNECTION_ERRORS` to a negative value removes basically all the disconnection logic from the built firmware. However, compared to running without this patch, that breaks the previously existing disconnect logic in matrix.c...

firetech commented:
> ```
> #broken
> handwired/lagrange:default
> whale/sk/v3:default
> ```

Both of these are using `SPLIT_TRANSPORT = custom`, so they're missing an implementation of `is_transport_connected`. A weak `return true;` implementation somewhere could solve this, unless we want to force custom transports to implement it.

firetech commented:
Used some tricks from tzarc's split disconnection branch to reduce firmware size a bit (fixed `helix/pico/sc:default` for me, but not the rest), and tweaked the possibilities to opt-out from this feature (can easily be changed to opt-in by changing default values).

There are now two ways to opt out:
* Set `SPLIT_MAX_CONNECTION_ERRORS` to 0. This will completely disable the connection status checks (also affects the slave matrix reset logic in matrix.c, though).
* Set `SPLIT_CONNECTION_CHECK_TIMEOUT` to 0. This will only disable the communication throttling while disconnected. Will make the firmware smaller (than without this patch, even).

drashna commented:
> Both of these are using `SPLIT_TRANSPORT = custom`, so they're missing an implementation of `is_transport_connected`. A weak `return true;` implementation somewhere could solve this, unless we want to force custom transports to implement it.

Ideally, it should, but for now, a weakly defined version should be okay, I think.

firetech commented:
It would technically also be possible to move this implementation to some other place (matrix.c? split_util.c?) where it could work for any transport (that returns false on error in `transport_master`). Would that be desirable?

firetech commented:
I made an alternate implementation compatible with custom transports: https://github.com/firetech/qmk_firmware/commit/split_common_connection_timeout_compat

Can include here if that's desired. I can see some benefits of having it limited to the standard transport, but it can be disabled with config regardless.

drashna commented:
`split_util.c` is probably the best place for it, and where I would have placed it. 👍🏻

drashna commented:
This has some merge conflicts now, sorry!

firetech commented:
> This has some merge conflicts now, sorry!

Should be fixed now. Thanks for the heads up. 🙂

firetech commented:
Lint is complaining about the lines that were part of the merge conflict... Should I change them back? 😜

drashna commented:
Welcome!  And yeah, been using this for a while now, and works well.

drashna commented:
> Lint is complaining about the lines that were part of the merge conflict... Should I change them back? 😜

Nah, it's fine.

firetech commented:
@KarlK90 Fair comments, but given the difference in size between the arrays, I (at least somewhat) feel that the current code is more readable (even though I _think_ your suggestions are correct). Also, I didn't write any of the lines you comment on (at most I moved them around a bit).

I might do some tests to see if/how it affects the performance (scan rate), though.

KarlK90 commented:
@firetech Sure, those are just some cleanup but I think it is nice to have. Regarding the speed-ups those could stem from the fact that ARM can do 32bit reads and compares if the length of the array matches, before it was comparing byte by byte.

firetech commented:
> Regarding the speed-ups those could stem from the fact that ARM can do 32bit reads and compares if the length of the array matches, before it was comparing byte by byte.

Hmm, in other words, the change to memcmp+memcpy for line 318-320 could hurt performance on AVR, since it could, in worst case, loop through the matrices twice, without any speed benefit?

KarlK90 commented:
~~Who cares about AVR~~

I think the compiler should identify that pattern and optimize it, but we should measure or look at the asm.

firetech commented:
The above force push was me reverting the memcmp/memcpy/memset change (due to deadline proximity), it will become a new PR together with a related change.

drashna commented:
A merge conflict here.

firetech commented:
> A merge conflict here.

Should be solved now. :)

zvecr commented on quantum/split_common/transport.c:
```diff
@@ -16,6 +16,7 @@
 
 #include <string.h>
 #include <debug.h>
+#include <timer.h>
```
Cant see any `timer_` call sites in thif file?
```suggestion
```

firetech replied:
Oops, that's a remnant from moving things around...
